pkg/kademlia/kademlia.go

```diff
@@ -120,7 +124,119 @@ func (k *Kademlia) GetRoutingTable(ctx context.Context) (dht.RoutingTable, error
 // Bootstrap contacts one of a set of pre defined trusted nodes on the network and
 // begins populating the local Kademlia node
 func (k *Kademlia) Bootstrap(ctx context.Context) error {
-	return k.dht.Bootstrap()
+	if len(k.bootstrapNodes) == 0 {
+		return BootstrapErr.New("no bootstrap nodes provided")
+	}
+	// psuedo randomly select a node from the slice of bootstrap nodes
+	// TODO(coyle): should we instead bootstrap against all the nodes ?
+	bn := k.bootstrapNodes[rand.Intn(len(k.bootstrapNodes))]
+
+	nodes, err := k.nodeClient.Lookup(ctx, bn, k.routingTable.Local())
+	if err != nil {
+		return BootstrapErr.Wrap(err)
+	}
+
+	if len(nodes) <= 0 {
+		return BootstrapErr.New("Bootstrap node provided no known nodes")
+	}
+
+	return k.boostrap(ctx, nodes, nil)
+}
+
+func (k *Kademlia) boostrap(ctx context.Context, nodes []*proto.Node, closestNode *proto.Node) error {
+	results, err := k.lookup(ctx, nodes)
+	if err != nil {
+		return err
+	}
+
+	closest := k.getClosest(results)
+	if cl := k.closer(&closest, closestNode); cl && closest.GetId() != closestNode.GetId() {
+		return k.boostrap(ctx, results, &closest)
+	}
+
+	_, err = k.lookup(ctx, results)
+
+	return err
+}
+
+func (k *Kademlia) lookup(ctx context.Context, nodes []*proto.Node) ([]*proto.Node, error) {
```

Comment:
a strategy that could work here -

 * someone calls k.Lookup(ctx, nodeID NodeID, k int) ([]*proto.Node, error)
 * k.Lookup creates a data structure of some kind with room for k nodes, and flags on whether or not those nodes have been talked to.
 * k.Lookup seeds the data structure with the k closest nodes in the local routing table, flagged that none of them have been contacted.
 * "alpha" worker goroutines are spun up concurrently. 
 * each goroutine is in a loop:
   - as long as there are uncontacted nodes in the nearest node data structure, 
   - mark the next closest node as in progress, do a lookup, and then get results
   - mark the node as complete
   - update the nearest-k data structure with the results of the request, bumping nodes out of the data structure that are farther than k.
   - start the loop over.
   - as soon as a goroutine discovers that all of the nodes in the list were either contacted or in progress, start a timer, 2x the time of the max response time so far. 
   - at the end of the timer cancel all remaining lookups
* return the nearest k successfully contacted nodes.